[Flang] Extracting internal constants from scalar literals

Leporacanthicus opened the pull request:
Constants actual arguments in function/subroutine calls are currently lowered as allocas + store. This can sometimes inhibit LTO and the constant will not be propagated to the called function. Particularly in cases where the function/subroutine call happens inside a condition.

This patch changes the lowering of these constant actual arguments to a global constant + fir.address_of_op. This lowering makes it easier for LTO to propagate the constant.

The optimization must be enabled explicitly to run.  Use -mmlir --enable-constant-argument-globalisation to enable.

Leporacanthicus committed:
[Flang] Extracting internal constants from scalar literals

Constants actual arguments in function/subroutine calls are currently lowered
as allocas + store. This can sometimes inhibit LTO and the constant will not
be propagated to the called function. Particularly in cases where the
function/subroutine call happens inside a condition.

This patch changes the lowering of these constant actual arguments to a
global constant + fir.address_of_op. This lowering makes it easier for
LTO to propagate the constant.

Co-authored-by: Dmitriy Smirnov <dmitriy.smirnov@arm.com>

kiranchandramohan reviewed (changes requested):
Thanks @Leporacanthicus. If I remember correctly, this pass requires the alias tags pass. Since that pass is reverted, this will have to wait.

And performance results have to be mentioned or present along with the patch.

vzakhari commented on flang/lib/Optimizer/Transforms/ConstExtruder.cpp:
```diff
@@ -0,0 +1,216 @@
+
+    patterns.insert<CallOpRewriter>(context, *di);
+    if (mlir::failed(
+            mlir::applyPartialConversion(func, target, std::move(patterns)))) {
```
If each call operation is expected to match/convert only once, you may consider using `applyPatternsAndFoldGreedily` with `GreedyRewriteStrictness::ExistingOps` option (and `enableRegionSimplification` set to false).

Leporacanthicus committed:
Use greedy rewriter

Leporacanthicus commented:
Benchmark for Spec-2017 503.bwaves is about 11% better on AArch64, and about 13% better on a fairly old x86-64 machine.

As far as I can tell, there's no slow-down, nor any improvement on other benchmarks.

tblah commented on flang/lib/Optimizer/Transforms/ConstExtruder.cpp:
```diff
@@ -0,0 +1,209 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "flang/Optimizer/Builder/BoxValue.h"
```
I don't think all of these includes are necessary

tblah commented on flang/test/Driver/bbc-mlir-pass-pipeline.f90:
```diff
@@ -31,6 +31,7 @@
 
 ! CHECK-NEXT: 'func.func' Pipeline
 ! CHECK-NEXT:   MemoryAllocationOpt
+! CHECK-NEXT:   ConstExtruderOpt
```
The indentation here makes this look like a func.func pass, which it is not. I think FileCheck does not check whitespace

tblah commented on flang/test/Fir/boxproc.fir:
```diff
@@ -16,9 +16,7 @@
 
 // CHECK-LABEL: define void @_QPtest_proc_dummy_other(ptr
 // CHECK-SAME:              %[[VAL_0:.*]])
-// CHECK:         %[[VAL_1:.*]] = alloca i32, i64 1, align 4
```
Could we accidentally drop alignment requirements when creating a global constant?

Leporacanthicus replied:
So, I don't think we have any control over alignment of global values. They are just "aligned as MLIR sees fit", which should work fine for align 4 on i32, but wouldn't work for an align 8 of i32, say. 

Do we even have access to alignment in FIR?  There's no mention in "FIRBuilder.cpp".

tblah replied:
Looking at it now, it is weird that this is written as `alloca` not `fir.alloca`. Is it from another dialect?

Which would explain how it has alignment, when I can't see a parameter for that in `fir.alloca`

kiranchandramohan replied:
@tblah This is an LLVM IR test.
```
// RUN: tco %s | FileCheck %s
```

tblah replied:
Ahh sorry for the noise. Maybe that test shouldn't be `.fir`, but anyway it wasn't created by this PR.

vzakhari commented on flang/lib/Optimizer/Transforms/ConstExtruder.cpp:
```diff
@@ -0,0 +1,209 @@
+#define DEBUG_TYPE "flang-const-extruder-opt"
+
+namespace {
+std::atomic<int> uniqueLitId = 1;
```
Can you please explain why it has to be atomic?

kiranchandramohan replied:
This could be a leftover from an initial version that Dmitriy wrote where this pass was not a module level pass.

kiranchandramohan commented on flang/lib/Optimizer/Transforms/ConstExtruder.cpp:
```diff
@@ -0,0 +1,209 @@
+          auto constant_def = store->getOperand(0).getDefiningOp();
+          // Expect constant definition operation
+          if (mlir::isa<mlir::arith::ConstantOp>(constant_def)) {
+            return true;
```
Should it be stricter, in the sense that there should only be one store that is a constant?

kiranchandramohan reviewed (commented):
A few minor comments.

Leporacanthicus commented on flang/lib/Optimizer/Transforms/ConstExtruder.cpp:
```diff
@@ -0,0 +1,209 @@
+
+  // is alloca
+  if (auto alloca = mlir::dyn_cast_or_null<fir::AllocaOp>(a->getDefiningOp())) {
+    // alloca has annotation
```
I've expanded the comment.

Leporacanthicus commented on flang/lib/Optimizer/Transforms/ConstExtruder.cpp:
```diff
@@ -0,0 +1,209 @@
+            globalName =
+                globalPrefix.str() + "." + std::to_string(uniqueLitId++);
+
+          if (alloca->hasOneUse()) {
```
And from what I have seen, this doesn't happen (at least not in the tests that fail if you bail out with when hasOneUse is false for the alloca.

Leporacanthicus commented on flang/lib/Optimizer/Transforms/ConstExtruder.cpp:
```diff
@@ -0,0 +1,209 @@
+            toErase.push_back(alloca);
+            toErase.push_back(stores[0]);
+          } else {
+            int count = -2;
```
Rewritten to start at zero and check for 2, which makes it more obvious what it does. I doubt it makes enough of a difference to obfuscate the code. [It would seem like something the compiler should be able to convert, anyway, if it's worth doing].

Leporacanthicus commented on flang/test/Transforms/const-extrude.f90:
```diff
@@ -0,0 +1,32 @@
+! RUN: %flang_fc1 -emit-fir %s -o - | fir-opt --const-extruder-opt | FileCheck %s
```
I will add a few more tests, and rewrite the original test as FIR-code rather than Fortran.

Leporacanthicus committed:
Fix review comments

Leporacanthicus committed:
Rename and add option to disable

tblah commented:
Thanks for the changes. LGTM once CI passes

Leporacanthicus committed:
Add more testing

Leporacanthicus committed:
Fix tests

Leporacanthicus committed:
Fix rebase issues

kiranchandramohan commented on flang/lib/Optimizer/Transforms/ConstantArgumentGlobalisation.cpp:
```diff
@@ -0,0 +1,190 @@
+//===- ConstExtruder.cpp --------------------------------------------------===//
```
Nit: Match file name.

kiranchandramohan reviewed (commented):
Nit comments to start the review.

tblah commented on flang/include/flang/Optimizer/Transforms/Passes.td:
```diff
@@ -252,6 +252,16 @@ def MemoryAllocationOpt : Pass<"memory-allocation-opt", "mlir::func::FuncOp"> {
+    Convert scalar literals of function arguments to global constants.
+  }];
+  let dependentDialects = [ "fir::FIROpsDialect" ];
+  let constructor = "::fir::createConstantArgumentGlobalisationPass()";
```
If you remove this line, tablegen will create a constructor for you. It will be called `createConstantArgumentGlobalisationOpt`.

tblah commented on flang/lib/Optimizer/Transforms/ConstantArgumentGlobalisation.cpp:
```diff
@@ -0,0 +1,190 @@
+          loc, newResultTypes,
+          callOp.getCallee().has_value() ? callOp.getCallee().value()
+                                         : mlir::SymbolRefAttr{},
+          newOperands, callOp.getFastmathAttr());
```
Is there a way to add *all* attributes from the old call to the new call? In case more attributes are added in the future

tblah commented on flang/lib/Optimizer/Transforms/ConstantArgumentGlobalisation.cpp:
```diff
@@ -0,0 +1,190 @@
+};
+
+// this pass attempts to convert immediate scalar literals in function calls
+// to global constants to allow transformations as Dead Argument Elimination
```
```suggestion
// to global constants to allow transformations such as Dead Argument Elimination
```

tblah commented on flang/lib/Optimizer/Transforms/ConstantArgumentGlobalisation.cpp:
```diff
@@ -0,0 +1,190 @@
+  void runOnOperation() override {
+    mlir::ModuleOp mod = getOperation();
+    mlir::DominanceInfo *di = &getAnalysis<mlir::DominanceInfo>();
+    mod.walk([di, this](mlir::func::FuncOp func) { runOnFunc(func, di); });
```
Please run this on other top level operations. It might be best to just run the CallOpRewriter on the whole module instead of on each function sequentially.

tblah commented on flang/lib/Optimizer/Transforms/ConstantArgumentGlobalisation.cpp:
```diff
@@ -0,0 +1,190 @@
+};
+} // namespace
+
+std::unique_ptr<mlir::Pass> fir::createConstantArgumentGlobalisationPass() {
+  return std::make_unique<ConstantArgumentGlobalisationOpt>();
+}
```
This can be generated automatically by tablegen

Leporacanthicus committed:
Review comment updates

kiranchandramohan reviewed (approved):
We probably have to document this a bit better.

Something like the following for the Pattern.
```
If the Call Ops argument is an alloca with `getAdaptToByRefAttrName` attribute.
If this alloca is used by a single Store and the store dominates the callOp and the store writes a constant to the alloca.

Additionally, delete the store if and only if the store dominates two operations (ie itself and the CallOp).
```

In general, the pass here works due to a mix of some guarantees provided by lowering and some enforced by the pass. It would be good to clarify what is guaranteed by each.

I was initially concerned that it works for the following function (where modify_me modifies the alloca value and print_me just prints it). I imagine such a situation will not arise in practise due to the lowering guarantees. 
```
func.func @func(%arg0: !fir.ref<i32>, %arg1: i1) {
  %c2_i32 = arith.constant 2 : i32
  %addr1 = fir.alloca i32 {adapt.valuebyref}
  fir.store %c2_i32 to %addr1 : !fir.ref<i32>
  fir.call @modify_me(%addr1) : (!fir.ref<i32>) -> ()
  fir.call @print_me(%addr1) : (!fir.ref<i32>) -> ()
  return
}
```

Should we consider adding this pass earlier in the flow?

Please wait for @vzakhari and @tblah.

Leporacanthicus committed:
Copy all attributes when creating new operation.

Instead of just copying the fast-math attribute, copy all of the
attributes of the call operation.

Also amend the test to check that attributes are still there in the
updated code.

Leporacanthicus commented on flang/lib/Optimizer/Transforms/ConstantArgumentGlobalisation.cpp:
```diff
@@ -0,0 +1,180 @@
+
+      // Delete if dominates itself and one more operation (which should
+      // be callOp)
+      if (count == 2)
+        toErase.push_back(store);
```
I do now: 
The call operation appears twice in the list of users. Not entirely surprising. Which means the count is 3. 

I'm not entirely sure what the solution is for this. 

I think perhaps counting AFTER the call has been updated, rather than before, would be a decent approach. I will try that and see what happens.

Leporacanthicus replied:
Well, that's not so easy, because we don't create the new operation until outside the loop. I'll take another look into that tomorrow. I think we should cope with this case.

Leporacanthicus replied:
Fixed with the latest update.

klausler commented:
The test program

```
subroutine s(x)
  print *, x
  x = x + 1.
end

call s(1.)
call s(1.)
end
```

will print `1.` twice successfully with default compilation options to five Fortran compilers today, and crash with a segfault with three.  this seems like the wrong thing to do for code portability.

Leporacanthicus commented:
> The test program
> 
> ```
> subroutine s(x)
>   print *, x
>   x = x + 1.
> end
> 
> call s(1.)
> call s(1.)
> end
> ```
> 
> will print `1.` twice successfully with default compilation options to five Fortran compilers today, and crash with a segfault with three. this seems like the wrong thing to do for code portability.

Yes, this is simpler variation on my simple assigned goto example that you said was "non-conforming", and I expect this one falls in the same category. It would probably be good to detect and warn for this - and I would expect at least some compilers do that.

klausler commented:
> > The test program
> > ```
> > subroutine s(x)
> >   print *, x
> >   x = x + 1.
> > end
> > 
> > call s(1.)
> > call s(1.)
> > end
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > will print `1.` twice successfully with default compilation options to five Fortran compilers today, and crash with a segfault with three. this seems like the wrong thing to do for code portability.
> 
> Yes, this is simpler variation on my simple assigned goto example that you said was "non-conforming", and I expect this one falls in the same category. It would probably be good to detect and warn for this - and I would expect at least some compilers do that.

Conforming or not, code that is reasonably portable today should also work with f18 with default options without source changes.  This case is kind of on the edge.

Leporacanthicus commented on flang/lib/Optimizer/Transforms/ConstantArgumentGlobalisation.cpp:
```diff
@@ -0,0 +1,180 @@
+      assert(!builder.getNamedGlobal(globalName) &&
+             "We should have a unique name here");
+
+      unsigned count = 0;
+      for (mlir::Operation *s : alloca->getUsers())
+        if (di.dominates(store, s))
+          ++count;
```
Also changed to cope with the test-case below, where the store wasn't being removed, so it's a bit different and in a different place now.

Leporacanthicus commented on flang/lib/Optimizer/Transforms/ConstantArgumentGlobalisation.cpp:
```diff
@@ -0,0 +1,180 @@
+            builder.create<fir::HasValueOp>(loc, val);
+          },
+          builder.createInternalLinkage());
+      mlir::Value addr = {builder.create<fir::AddrOfOp>(
+          loc, global.resultType(), global.getSymbol())};
```
No idea, probably leftover from some more complex code, is my best guess.

Leporacanthicus commented:
> Conforming or not, code that is reasonably portable today should also work with f18 with default options without source changes. This case is kind of on the edge.

This particular example works fine (as in prints 1 twice) with these changes - because the write to x gets optimized away, I think. 

I'm not sure if there are more complex cases where it doesn't work quite that well - I have a hard time coming up with a case where this pass actually goes wrong. Your comments makes me wonder what we should do for that - the only plausible solution I can think of is to basically not do this unless we the compiler can "see" the source of the code being called - and it's precisely the case of the callee being in a different module that we're trying to make better here.

Any thoughts?

Leporacanthicus committed:
Fix further review comments.

Additionally fixed a bug where same argument used twice would not have the
store and alloca removed.

Add test to check for those instructions being removed.

Leporacanthicus commented:
> > Conforming or not, code that is reasonably portable today should also work with f18 with default options without source changes. This case is kind of on the edge.
> 
> This particular example works fine (as in prints 1 twice) with these changes - because the write to x gets optimized away, I think.
> 
> I'm not sure if there are more complex cases where it doesn't work quite that well - I have a hard time coming up with a case where this pass actually goes wrong. Your comments makes me wonder what we should do for that - the only plausible solution I can think of is to basically not do this unless we the compiler can "see" the source of the code being called - and it's precisely the case of the callee being in a different module that we're trying to make better here.
> 
> Any thoughts?

@klausler : Would it help if we only enable this on -O2 or something like that? Then it's not enabled without the user specifying a higher optimiseation level.

klausler commented:
> > > Conforming or not, code that is reasonably portable today should also work with f18 with default options without source changes. This case is kind of on the edge.
> > 
> > 
> > This particular example works fine (as in prints 1 twice) with these changes - because the write to x gets optimized away, I think.
> > I'm not sure if there are more complex cases where it doesn't work quite that well - I have a hard time coming up with a case where this pass actually goes wrong. Your comments makes me wonder what we should do for that - the only plausible solution I can think of is to basically not do this unless we the compiler can "see" the source of the code being called - and it's precisely the case of the callee being in a different module that we're trying to make better here.
> > Any thoughts?
> 
> @klausler : Would it help if we only enable this on -O2 or something like that? Then it's not enabled without the user specifying a higher optimiseation level.

It seems weird to have an optimization level imply less portable semantics that are needed to work around a weakness in an optimization pass.  Would it not be better to improve LTO's handling of constant arguments passed by reference?

Leporacanthicus committed:
Make constant argument globalisation disabled by default.

This pass can convert local constant arguments into global values, but it can
have negative effects on code that is borderline to non-conforming, where a
constant argument is modified inside the callee - many compilers accept such
code. So, to use this pass, it needs to be enabled explicitly.

To enable in flang-new, use -mmlir --enable-constant-argument-globalisation

Leporacanthicus commented:
> > @klausler : Would it help if we only enable this on -O2 or something like that? Then it's not enabled without the user specifying a higher optimiseation level.
> 
> It seems weird to have an optimization level imply less portable semantics that are needed to work around a weakness in an optimization pass. Would it not be better to improve LTO's handling of constant arguments passed by reference?

I can see that point. I've changed it so that you have to manually enable this particular pass.

Leporacanthicus committed:
Rebase

Leporacanthicus committed:
Fix formatting

Leporacanthicus committed:
Fix double erase issue

Leporacanthicus commented:
@klausler Are you OK with this going in disabled? We have no plans to enable it generically, but we'd like to have it so we can argue for "it makes this much difference to <benchmark>, so if we can get LLVM to do this, it would be good".

Leporacanthicus committed:
Fix trivial typo

Leporacanthicus commented on flang/include/flang/Tools/CLOptions.inc:
```diff
@@ -86,8 +90,8 @@ DisableOption(BoxedProcedureRewrite, "boxed-procedure-rewrite",
-DisableOption(ConstantArgumentGlobalisation, "constant-argument-globalisation",
-    "disable the local constants to global constant conversion");
+EnableOption(ConstantArgumentGlobalisation, "constant-argument-globalisation",
+    "enable the local constants to global constant conversion");
```
Good spot. It was wrong in the old code too, it adds "disable" automatically, so it had double disable too.
